types/wlr_surface.c

```diff
@@ -255,6 +261,7 @@ static void surface_state_copy(struct wlr_surface_state *state,
 		struct wlr_surface_state *next) {
 	state->width = next->width;
 	state->height = next->height;
+	state->buffer = next->buffer;
```

Comment:
You might notice that `surface_state_copy` is a bit weird: the `buffer_resource` doesn't get copied.

The reason is we have `wlr_surface.previous` which holds the state for the _previous_ commit. We use `surface_state_copy` to construct `previous`. However we don't want to populate the buffer fields of the previous state, because that would lock the buffer for longer than necessary (and force the client to allocate and use one more buffer).

I don't like `wlr_surface.previous`, I think we should get rid of it, but that's another battle (a few more details about this in https://github.com/swaywm/wlroots/issues/2098). In the meantime, move this logic into `surface_state_move` (where you can notice `buffer_resource` is handled as well).